### PR TITLE
Allow attaching service pods to additional network

### PR DESCRIPTION
**Pull request by stuggi**
For network isolation outgoing traffic, service pods need to be attached to the network used by the edpm nodes. To achive this, the `k8s.v1.cni.cncf.io/networks` annotation needs to be set on the pods. This adds a networkAttachment parameter to the CRD where it can be specified to which additional network the service pods should be attached to.

```
spec:
...
      networkAttachment: internalapi
```

Depends-On: openstack-k8s-operators/lib-common#133

### CODE REVIEW

**Comment by stuggi** [comment]
for testing quay.io/mschuppe/openstack-operator-index:2cbfda9fc2d54837ef9e4e361402096736b38669 could be used which has references to the required ovs/ovn and also edpm operators.

File: pkg/ovs/daemonset.go [hunk 1]
```diff
@@ -78,16 +79,15 @@ func DaemonSet(
 	envVars["OvnBridge"] = env.SetValue(instance.Spec.ExternalIDS.OvnBridge)
 	envVars["OvnRemote"] = env.SetValue(dbmap["SB"])
 	envVars["OvnEncapType"] = env.SetValue(instance.Spec.ExternalIDS.OvnEncapType)
+	// TODO
+	envVars["PodNamespace"] = env.SetValue(instance.Namespace)
+	envVars["PodNetworksStatus"] = EnvDownwardAPI("metadata.annotations['k8s.v1.cni.cncf.io/networks-status']")
+	envVars["OvnEncapNetwork"] = env.SetValue("internalapi")
```
Comment (stuggi) [inline pkg/ovs/daemonset.go:85]: this needs update to not be fixed for sure and get it from the NetworkAttachment

File: api/bases/ovs.openstack.org_ovs.yaml [hunk 2]
```diff
@@ -56,6 +56,11 @@ spec:
                 - ovn-bridge
                 - system-id
                 type: object
+              networkAttachment:
```
Comment (stuggi) [inline api/bases/ovs.openstack.org_ovs.yaml:59]: would there be a uses case to attach the pods to multiple networks and have another param to specify the one which references the network from which the interface IP should be used for the OvnEncapIP?

File: templates/ovs/bin/get_net_ip [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+def get_ip_from_network_status():
+    namespace = os.environ.get('PodNamespace')
+    net_status = json.loads(os.environ.get('PodNetworksStatus'))
+    net_attachment_name = sys.argv[1]
+
+    for net in net_status:
+        if net.get('name') == '/'.join([namespace, net_attachment_name]):
+            ips = net.get('ips', [])
+            if len(ips) > 0:
```
Comment (booxter) [inline templates/ovs/bin/get_net_ip:49]: nit: just replace with "if ips", we don't care about the exact length of the list, just that it's not empty

File: pkg/ovs/daemonset.go [hunk 4]
```diff
@@ -111,16 +108,10 @@ func DaemonSet(
 						{
 							Name: ServiceName + "db-server",
 							Command: []string{
-								"/usr/local/bin/container-scripts/start-ovsdb-server.sh",
+								"/usr/bin/start-ovs",
```
Comment (booxter) [inline pkg/ovs/daemonset.go:111]: where do you get it from? the official image we use for the container - openstack-ovn-base - doesn't contain the script. Hence the need to drive the startup of services via a shell script.

**Review by booxter — CHANGES_REQUESTED** [review]
I have a question about the changes in ovsdb-server startup (switching to start-ovs script). I believe we no longer carry the script in the images used - which are tripleo openstack-ovn-base, not the custom image from slaweq that was used in the past.

File: pkg/ovs/daemonset.go (hunk 4 above)
Comment (stuggi) [reply pkg/ovs/daemonset.go:111]: I did not plan to update this, must be a bad rebase. will fix it

**Review by stuggi — COMMENTED** [review]
/hold

**Review by booxter — APPROVED** [review]

File: api/bases/ovs.openstack.org_ovs.yaml (hunk 2 above)
Comment (slawqo) [reply api/bases/ovs.openstack.org_ovs.yaml:59]: sorry for being late into this but I think that there is such use case as this pod should be connected usually to "internalapi" and to "tenant" networks for tunnels. And that "tenant" network should be used as the one from which OvnEncapIP is taken. I will propose update for that to the ovs-operator

Comment (slawqo) [reply api/bases/ovs.openstack.org_ovs.yaml:59]: Actually no, it needs to connect to ovn central pods but it can use "pods" network probably, right?

**Pull request merged**